netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -225,18 +265,17 @@ public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exce
   public void channelInactive(ChannelHandlerContext ctx) throws Exception {
     try {
       logger.fine("Network channel is closed");
+      // Set or update the GOAWAY status to indicate the connection has broken.
       goAwayStatus(goAwayStatus().augmentDescription("Network channel closed"));
+
+      // Cancel any pending ping request.
       cancelPing();
-      // Report status to the application layer for any open streams
-      connection().forEachActiveStream(new Http2StreamVisitor() {
-        @Override
-        public boolean visit(Http2Stream stream) throws Http2Exception {
-          clientStream(stream).transportReportStatus(goAwayStatus, false, new Metadata.Trailers());
-          return true;
-        }
-      });
+
+      // Close the buffered encoder and cancel any pending streams.
+      encoder().close();
```

Comment:
This is the main thrust of this PR. We're clearing out the buffering encoder's cached streams before we delegate to the base class to close any active streams.  Otherwise the buffering encoder has to rely on repeated callbacks from stream closures to clear itself out one stream at a time.